File: common/src/main/java/org/valkyrienskies/create_interactive/mixinducks/ContraptionDuck.java
package org.valkyrienskies.create_interactive.mixinducks;

import com.simibubi.create.content.contraptions.behaviour.MovementContext;
import net.minecraft.core.BlockPos;
import net.minecraft.world.level.levelgen.structure.templatesystem.StructureTemplate;
import org.apache.commons.lang3.tuple.Pair;
import org.jetbrains.annotations.Nullable;

import java.util.Collection;

public interface ContraptionDuck {
    void ci$setBlock(BlockPos localPos, StructureTemplate.StructureBlockInfo structureBlockInfo);

    @SuppressWarnings("unused")
    boolean ci$hasActorAtPos(BlockPos localPos);

    @SuppressWarnings("unused")
    boolean ci$hasBogeyAtPos(BlockPos localPos);

    @SuppressWarnings("unused")
    @Nullable
    Pair<StructureTemplate.StructureBlockInfo, MovementContext> ci$getActorAtPos(BlockPos localPos);

    @SuppressWarnings("unused")
    Collection<BlockPos> ci$getChangedActors();

    @SuppressWarnings("unused")
    void ci$clearChangedActors();
}

File: common/src/main/kotlin/org/valkyrienskies/create_interactive/mixinducks/ContraptionDuck.kt
package org.valkyrienskies.create_interactive.mixinducks

import net.minecraft.core.BlockPos
import net.minecraft.world.level.levelgen.structure.templatesystem.StructureTemplate
import org.valkyrienskies.create_interactive.CreateActorImmutable

interface ContraptionDuck {
    fun `ci$setBlock`(localPos: BlockPos, structureBlockInfo: StructureTemplate.StructureBlockInfo)

    fun `ci$hasActorAtPos`(localPos: BlockPos): Boolean

    fun `ci$hasBogeyAtPos`(localPos: BlockPos): Boolean

    fun `ci$getActorAtPos`(localPos: BlockPos): CreateActorImmutable?

    fun `ci$getChangedActors`(): Collection<BlockPos>

    fun `ci$clearChangedActors`()
}
